From 8ea0037f4b88c83013cc2636e48ccbac24c9ec87 Mon Sep 17 00:00:00 2001
From: honjow <honjow311@gmail.com>
Date: Thu, 6 Mar 2025 17:35:50 +0800
Subject: [PATCH 1/6] Add Claw 8

---
 msi-ec.c | 68 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 68 insertions(+)

diff --git a/msi-ec.c b/msi-ec.c
index 4d55ebb..52a1c57 100644
--- a/msi-ec.c
+++ b/msi-ec.c
@@ -2886,6 +2886,73 @@ static struct msi_ec_conf CONF40 __initdata = {
 	},
 };
 
+static const char *ALLOWED_FW_401[] __initconst = {
+	"1T52EMS1.104", // MSI Claw 8 AI+ A2VM
+	NULL
+};
+
+static struct msi_ec_conf CONF401 __initdata = {
+	.allowed_fw = ALLOWED_FW_401,
+	.charge_control_address = 0xd7,
+	.webcam = {
+		.address       = MSI_EC_ADDR_UNSUPP,
+		.block_address = 0x2f,
+		.bit           = 1,
+	},
+	.fn_win_swap = {
+		.address = MSI_EC_ADDR_UNSUPP,
+		.bit     = 4,
+		.invert  = false,
+	},
+	.cooler_boost = {
+		.address = 0x98,
+		.bit     = 7,
+	},
+	.shift_mode = {
+		.address = 0xd2,
+		.modes = {
+			{ SM_ECO_NAME,     0xc2 },
+			{ SM_COMFORT_NAME, 0xc1 },
+			{ SM_SPORT_NAME,   0xc0 },
+			MSI_EC_MODE_NULL
+		},
+	},
+	.super_battery = {
+		.address = MSI_EC_ADDR_UNSUPP,
+		.mask    = 0x0f,
+	},
+	.fan_mode = {
+		.address = 0xd4,
+		.modes = {
+			{ FM_AUTO_NAME,     0x0d },
+			{ FM_SILENT_NAME,   0x1d },
+			{ FM_ADVANCED_NAME, 0x8d },
+			MSI_EC_MODE_NULL
+		},
+	},
+	.cpu = {
+		.rt_temp_address       = 0x68,
+		.rt_fan_speed_address  = 0x71,
+	},
+	.gpu = {
+		.rt_temp_address      = 0x80,
+		.rt_fan_speed_address = 0x89,
+	},
+	.leds = {
+		.micmute_led_address = MSI_EC_ADDR_UNSUPP,
+		.mute_led_address    = MSI_EC_ADDR_UNSUPP,
+		.bit                 = 1,
+	},
+	.kbd_bl = {
+		.bl_mode_address  = MSI_EC_ADDR_UNSUPP, // KB auto turn off
+		.bl_modes         = { 0x00, 0x08 }, // always on; off after 10 sec
+		.max_mode         = 1,
+		.bl_state_address = MSI_EC_ADDR_UNSUPP,
+		.state_base_value = 0x80,
+		.max_state        = 3,
+	},
+};
+
 static const char *ALLOWED_FW_41[] __initconst = {
 	"15M1IMS2.111", // MSI Vector 16 HX A14VHG
 	NULL
@@ -3060,6 +3127,7 @@ static struct msi_ec_conf *CONFIGURATIONS[] __initdata = {
 	&CONF38,
 	&CONF39,
 	&CONF40,
+	&CONF401,
 	&CONF41,
 	&CONF42,
 	NULL
-- 
2.48.1

